// slnodejs/tsOutputs/build-scanner/js_diff/source-files-notifier.js
var CockpitNotifier = require("../../common/agent-events/cockpit-notifier").CockpitNotifier;
var troubleshootingPage = "https://sealights.atlassian.net/wiki/spaces/SUP/pages/1172144129/Javascript+and+Node.js+Source+Map+Files";
function SourceFilesNotifier(logger) {
    this.MAX_FILES_NO_NOTIFY = 10;
    this.counter = 0;
    this.logger = logger;
}
SourceFilesNotifier.prototype.notifySources = function (sourcesArr, generatedFile, hasColumnInfo) {
    if (this.counter < this.MAX_FILES_NO_NOTIFY) {
        hasColumnInfo ? notifySourceWithColumn(sourcesArr, generatedFile) : warnSourceWithOutColumn(sourcesArr, generatedFile);
        this.counter++;
    }
    else {
        this.logger.debug(`Not sending sources for file '${generatedFile}'. already sent sources for '${this.MAX_FILES_NO_NOTIFY}' files`);
    }
};
function notifySourceWithColumn(sourcesArr, generatedFile) {
    var message = JSON.stringify({
        generatedFileName: generatedFile,
        sources: sourcesArr
    });
    CockpitNotifier.sendGenericMessage(message);
}
function warnSourceWithOutColumn(sourcesArr, generatedFile) {
    var message = "Detected source-maps without column information, missing column information may result in inaccurate coverage. " +
        "For more information: " + troubleshootingPage +
        ". Generated file: " + generatedFile +
        ". Sources: " + sourcesArr;
    CockpitNotifier.sendWarning(message);
}
module.exports = SourceFilesNotifier;
