// slnodejs/tsOutputs/cli-parse/executors/mocha-executor.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MochaExecutor = void 0;
const test_listener_executor_1 = require("./test-listener-executor");
const test_runner_build_session_data_builder_1 = require("../../common/utils/test-runner-build-session-data-builder");
const child_process_wrapper_1 = require("../child-process-wrapper");
const path_1 = require("path");
const agent_loader_1 = require("../../test-listener/agent-loader");
class MochaExecutor extends test_listener_executor_1.TestListenerExecutor {
    constructor() {
        super(...arguments);
        this.mochaReporterPath = (0, path_1.resolve)(__dirname, '../../mocha-reporter/index');
        this.commandName = 'mocha';
    }
    innerExecute(args, extraParams) {
        const _super = Object.create(null, {
            innerExecute: { get: () => super.innerExecute }
        });
        return __awaiter(this, void 0, void 0, function* () {
            yield _super.innerExecute.call(this, args, extraParams);
            const mochaPath = this.getMochaPath();
            let mochaParams = [require.resolve(mochaPath)];
            if (this.useIstanbul == true) {
                mochaParams.push('--');
            }
            mochaParams.push('-R');
            mochaParams.push(this.mochaReporterPath);
            mochaParams = mochaParams.concat(extraParams);
            agent_loader_1.AgentLoader.run(mochaParams, this.useIstanbul, this.spawnRuntimeArgs, false, {
                command: this.commandName
            });
        });
    }
    getBuildSessionData(backendProxy, cmd) {
        return __awaiter(this, void 0, void 0, function* () {
            const buildSessionDataBuilder = new test_runner_build_session_data_builder_1.TestRunnerBuildSessionDataBuilder(backendProxy);
            return yield buildSessionDataBuilder.build(cmd);
        });
    }
    runWithoutSealights() {
        return __awaiter(this, void 0, void 0, function* () {
            const mochaPath = require.resolve(this.getMochaPath());
            const allArgs = this.extraArgs;
            return child_process_wrapper_1.ChildProcessWrapper.runCommand(mochaPath, allArgs);
        });
    }
    getMochaPath() {
        return /^win/.test(process.platform) ? 'mocha/bin/_mocha' : '.bin/_mocha';
    }
}
exports.MochaExecutor = MochaExecutor;
