// sealights-cypress-plugin/tsOutputs/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TEST_STATUS = exports.SL_COLORING_HEADERS = exports.SL_COLORING_EVENTS = exports.SL_ERROR_LOG = exports.SL_DEBUG_LOG = exports.SL_SEND_COVERAGE = exports.SL_GET_TEST_SESSION_ID = exports.SL_REPORT_SPEC_END_TASK = exports.SL_REPORT_TEST_END_TASK = exports.X_SL_SERVER = exports.SL_TASKS_REGISTERED = exports.SL_END_TEST_SESSION = exports.SL_START_TEST_SESSION = exports.SL_CYPRESS_ENVIRONMENT = void 0;
exports.SL_CYPRESS_ENVIRONMENT = {
    BUILD_SESSION_ID: 'SL_BUILD_SESSION_ID',
    TOKEN: 'SL_TOKEN',
    TEST_STAGE: 'SL_TEST_STAGE',
    LAB_ID: 'SL_LAB_ID',
    ENABLE_REMOTE_AGENT: 'SL_ENABLE_REMOTE_AGENT',
    PER_FILE_REPORTING: 'SL_PER_FILE_REPORTING',
    PROXY: 'SL_PROXY',
    COLLECTOR_URL: 'SL_COLLECTOR_URL',
    DROP_FOOTPRINTS: 'SL_DROP_FOOTPRINTS',
    ENABLE_PLUGINS_API: 'SL_ENABLE_PLUGINS_API',
    LITE_MODE: 'SL_LITE_MODE',
    LITE_MODE_MANAGE_EXECUTIONS: 'SL_LITE_MODE_MANAGE_EXECUTIONS',
    TIA_DISABLED: 'SL_TIA_DISABLED',
};
exports.SL_START_TEST_SESSION = 'sealightsStartSession';
exports.SL_END_TEST_SESSION = 'sealightsEndSession';
exports.SL_TASKS_REGISTERED = 'sealightsTasksRegistered';
exports.X_SL_SERVER = 'x-sl-server';
exports.SL_REPORT_TEST_END_TASK = 'reportSealightsTestEnd';
exports.SL_REPORT_SPEC_END_TASK = 'reportSealightsSpecEnd';
exports.SL_GET_TEST_SESSION_ID = 'getSealightsTestSession';
exports.SL_SEND_COVERAGE = 'reportCoverage';
exports.SL_DEBUG_LOG = 'slDebug';
exports.SL_ERROR_LOG = 'slError';
exports.SL_COLORING_EVENTS = {
    SET_CONTEXT: 'set:context',
    DELETE_CONTEXT: 'delete:context',
};
exports.SL_COLORING_HEADERS = {
    TEST_NAME: 'x-sl-test-name',
    TEST_SESSION_ID: 'x-sl-test-session-id',
};
var TEST_STATUS;
(function (TEST_STATUS) {
    TEST_STATUS["PASSED"] = "passed";
    TEST_STATUS["SKIPPED"] = "skipped";
    TEST_STATUS["FAILED"] = "failed";
})(TEST_STATUS || (exports.TEST_STATUS = TEST_STATUS = {}));
