// slnodejs/tsOutputs/common/agent-events/agent-events-contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentTechnologies = exports.AgentEventCode = exports.AgentTypes = void 0;
var AgentTypes;
(function (AgentTypes) {
    AgentTypes["BUILD_SCANNER"] = "BuildScanner";
    AgentTypes["TEST_LISTENER"] = "TestListener";
    AgentTypes["BROWSER_AGENT"] = "BrowserAgent";
    AgentTypes["SLNODEJS"] = "slnodejs";
})(AgentTypes = exports.AgentTypes || (exports.AgentTypes = {}));
var AgentEventCode;
(function (AgentEventCode) {
    // Events
    AgentEventCode[AgentEventCode["GENERIC_AGENT_EVENT"] = 1000] = "GENERIC_AGENT_EVENT";
    AgentEventCode[AgentEventCode["AGENT_STARTED"] = 1001] = "AGENT_STARTED";
    AgentEventCode[AgentEventCode["AGENT_SHUTDOWN"] = 1002] = "AGENT_SHUTDOWN";
    AgentEventCode[AgentEventCode["AGENT_PING"] = 1003] = "AGENT_PING";
    AgentEventCode[AgentEventCode["AGENT_CONFIG_CHANGED"] = 1004] = "AGENT_CONFIG_CHANGED";
    AgentEventCode[AgentEventCode["FIRST_COVERAGE_INSTRUMENTATION_PERFORMED"] = 1005] = "FIRST_COVERAGE_INSTRUMENTATION_PERFORMED";
    AgentEventCode[AgentEventCode["FIRST_TIME_HAS_EXECUTION"] = 1006] = "FIRST_TIME_HAS_EXECUTION";
    AgentEventCode[AgentEventCode["FIRST_TIME_NO_EXECUTION"] = 1007] = "FIRST_TIME_NO_EXECUTION";
    AgentEventCode[AgentEventCode["AGENT_MUTED"] = 1008] = "AGENT_MUTED";
    AgentEventCode[AgentEventCode["AGENT_UNMUTED"] = 1009] = "AGENT_UNMUTED";
    AgentEventCode[AgentEventCode["FIRST_TIME_COLLECTED_FP"] = 1010] = "FIRST_TIME_COLLECTED_FP";
    AgentEventCode[AgentEventCode["CONTEXT_PROPAGATION_TELEMETRY"] = 1019] = "CONTEXT_PROPAGATION_TELEMETRY";
    // Info
    AgentEventCode[AgentEventCode["GENERIC_MESSAGE"] = 2000] = "GENERIC_MESSAGE";
    AgentEventCode[AgentEventCode["GENERIC_MESSAGE_SUPERUSER"] = 2999] = "GENERIC_MESSAGE_SUPERUSER";
    // Warnings
    AgentEventCode[AgentEventCode["WARN"] = 3000] = "WARN";
    AgentEventCode[AgentEventCode["AGENT_DID_NOT_SHUTDOWN"] = 3001] = "AGENT_DID_NOT_SHUTDOWN";
    AgentEventCode[AgentEventCode["GENERIC_WARNING_SUPERUSER"] = 3999] = "GENERIC_WARNING_SUPERUSER";
    AgentEventCode[AgentEventCode["GIT_SUBMODULES_DETECTED"] = 3501] = "GIT_SUBMODULES_DETECTED";
    // Errors
    AgentEventCode[AgentEventCode["GENERIC_ERROR"] = 4000] = "GENERIC_ERROR";
    AgentEventCode[AgentEventCode["DUPLICATE_MODULE"] = 4001] = "DUPLICATE_MODULE";
    AgentEventCode[AgentEventCode["DATA_PROCESSOR_NO_EXECUTIONS"] = 4002] = "DATA_PROCESSOR_NO_EXECUTIONS";
    AgentEventCode[AgentEventCode["DATA_PROCESSOR_INVALID_FORMAT"] = 4003] = "DATA_PROCESSOR_INVALID_FORMAT";
    AgentEventCode[AgentEventCode["DATA_PROCESSOR_EMPTY_DATA"] = 4004] = "DATA_PROCESSOR_EMPTY_DATA";
    AgentEventCode[AgentEventCode["BUILD_MAP_SUBMISSION_ERROR"] = 4005] = "BUILD_MAP_SUBMISSION_ERROR";
    AgentEventCode[AgentEventCode["FOOTPRINTS_SUBMISSION_ERROR"] = 4006] = "FOOTPRINTS_SUBMISSION_ERROR";
    AgentEventCode[AgentEventCode["TEST_EVENTS_SUBMISSION_ERROR"] = 4007] = "TEST_EVENTS_SUBMISSION_ERROR";
    AgentEventCode[AgentEventCode["EXTERNAL_DATA_PROCESSOR_SUBMISSION_ERROR"] = 4008] = "EXTERNAL_DATA_PROCESSOR_SUBMISSION_ERROR";
    AgentEventCode[AgentEventCode["UNSUPPORTED_OS"] = 4009] = "UNSUPPORTED_OS";
    AgentEventCode[AgentEventCode["UNSUPPORTED_RUNTIME"] = 4010] = "UNSUPPORTED_RUNTIME";
    AgentEventCode[AgentEventCode["THIRD_PARTY_PACKAGE_DETECTED"] = 4011] = "THIRD_PARTY_PACKAGE_DETECTED";
    AgentEventCode[AgentEventCode["THIRD_PARTY_FILE_DETECTED"] = 4012] = "THIRD_PARTY_FILE_DETECTED";
    AgentEventCode[AgentEventCode["OTEL_ERROR"] = 4013] = "OTEL_ERROR";
    AgentEventCode[AgentEventCode["STATIC_INSTRUMENTATION_ERROR"] = 4021] = "STATIC_INSTRUMENTATION_ERROR";
    AgentEventCode[AgentEventCode["FOOTPRINTS_LOSS"] = 4999] = "FOOTPRINTS_LOSS";
    // Logs
    AgentEventCode[AgentEventCode["LEAST_VERBOSE_LOG"] = 5001] = "LEAST_VERBOSE_LOG";
    AgentEventCode[AgentEventCode["MOST_VERBOSE_LOG"] = 5999] = "MOST_VERBOSE_LOG";
})(AgentEventCode = exports.AgentEventCode || (exports.AgentEventCode = {}));
var AgentTechnologies;
(function (AgentTechnologies) {
    AgentTechnologies["NODEJS"] = "nodejs";
    AgentTechnologies["BROWSER"] = "browser";
})(AgentTechnologies = exports.AgentTechnologies || (exports.AgentTechnologies = {}));
