// slnodejs/tsOutputs/common/agent-events/agent-start-info-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentStartInfoBuilder = void 0;
const data_cleansing_utils_1 = require("../utils/data-cleansing-utils");
class AgentStartInfoBuilder {
    constructor(agentInstanceInfoBuilder, machineInfoBuilder, techSpecificInfoBuilder, ciInfoBuilder, agentConfig) {
        this.agentInstanceInfoBuilder = agentInstanceInfoBuilder;
        this.machineInfoBuilder = machineInfoBuilder;
        this.techSpecificInfoBuilder = techSpecificInfoBuilder;
        this.ciInfoBuilder = ciInfoBuilder;
        this.agentConfig = agentConfig;
        this.agentInstanceInfoBuilder = agentInstanceInfoBuilder;
        this.machineInfoBuilder = machineInfoBuilder;
        this.techSpecificInfoBuilder = techSpecificInfoBuilder;
        this.ciInfoBuilder = ciInfoBuilder;
    }
    fillData() {
    }
    build() {
        const response = {
            agentInfo: this.agentInstanceInfoBuilder.build(),
            machineInfo: this.machineInfoBuilder.build(),
            techSpecificInfo: this.techSpecificInfoBuilder.build(),
            ciInfo: this.ciInfoBuilder.build()
        };
        if (this.agentConfig.removeSensitiveData.value)
            return data_cleansing_utils_1.DataCleansingUtils.removeSensitiveData(response);
        return response;
    }
}
exports.AgentStartInfoBuilder = AgentStartInfoBuilder;
