// slnodejs/tsOutputs/common/agent-events/no-op-agent-events-controller.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoOpAgentEventsController = void 0;
const agent_events_controller_1 = require("./agent-events-controller");
class NoOpAgentEventsController extends agent_events_controller_1.AgentEventsController {
    submitAgentStartedEvent(packageJsonFile) {
        return __awaiter(this, void 0, void 0, function* () {
        });
    }
    submitAgentEventRequest(event) {
        return __awaiter(this, void 0, void 0, function* () {
            return true;
        });
    }
    sendMessage(message, code) {
    }
    submitAgentShutdownEvent() {
        return __awaiter(this, void 0, void 0, function* () {
            return null;
        });
    }
    submitPingEvent() {
    }
    submitEvent(code, data) {
    }
    submitEventOnce(code) {
    }
    submitGenericMessage(message) {
    }
    submitWarning(message) {
    }
    submitError(message) {
    }
    submitErrorsBatch(messages) {
    }
}
exports.NoOpAgentEventsController = NoOpAgentEventsController;
