// slnodejs/tsOutputs/cli-parse/contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMMAND_ARGS = exports.COMMANDS = void 0;
var COMMANDS;
(function (COMMANDS) {
    COMMANDS["CONFIG"] = "config";
    COMMANDS["PR_CONFIG"] = "prConfig";
    COMMANDS["BUILD"] = "build";
    COMMANDS["SCAN"] = "scan";
    COMMANDS["MOCHA"] = "mocha";
    COMMANDS["JASMINE"] = "jasmine";
    COMMANDS["MOCHA_PHANTOM_JS"] = "mochaPhantomjs";
    COMMANDS["RUN"] = "run";
    COMMANDS["START"] = "start";
    COMMANDS["END"] = "end";
    COMMANDS["UPLOAD_REPORTS"] = "uploadReports";
    COMMANDS["EXTERNAL_REPORT"] = "externalReport";
    COMMANDS["NYC_REPORT"] = "nycReport";
    COMMANDS["INSTALL_BABEL"] = "installBabel";
    COMMANDS["COMPONENT_UPDATE"] = "componentUpdate";
    COMMANDS["COMPONENT_DELETE"] = "componentDelete";
    COMMANDS["BUILD_END"] = "buildend";
    COMMANDS["SCANNED"] = "scanned";
    COMMANDS["DRY_RUN"] = "dryRun";
})(COMMANDS = exports.COMMANDS || (exports.COMMANDS = {}));
var COMMAND_ARGS;
(function (COMMAND_ARGS) {
    COMMAND_ARGS["USE_OTEL"] = "useOtel";
    COMMAND_ARGS["ENABLE_OPEN_TELEMETRY"] = "enableOpenTelemetry";
    COMMAND_ARGS["INSTRUMENT_FOR_BROWSERS"] = "instrumentForBrowsers";
})(COMMAND_ARGS = exports.COMMAND_ARGS || (exports.COMMAND_ARGS = {}));
