// slnodejs/tsOutputs/build-scanner/instrumentation/content-instrumenter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentInstrumenter = void 0;
const libInstrument = require("istanbul-lib-instrument");
const slLibInstrument = require("sl-istanbul-lib-instrument");
const instrumentation_utils_1 = require("./instrumentation-utils");
class ContentInstrumenter {
    constructor(config) {
        const instrumenterConfig = {
            coverageVariable: '$SealightsCoverage_' + config.buildSessionId,
            embedSource: false,
            compact: true,
            preserveComments: true,
            esModules: config.esModules,
            instrumentLineLevel: config.instrumentLineLevel
        };
        this.instrumenter = config.newInstrumenter ?
            slLibInstrument.createInstrumenter(instrumenterConfig) :
            libInstrument.createInstrumenter(instrumenterConfig);
    }
    instrumentContent(content, fileKey) {
        content = (0, instrumentation_utils_1.removeWebpackComments)(content);
        let instrumentedContent = this.instrumenter.instrumentSync(content, fileKey);
        instrumentedContent = (0, instrumentation_utils_1.unescapeUnicode0275)(instrumentedContent);
        instrumentedContent = (0, instrumentation_utils_1.fixMissingSpaceAfterReturnStatement)(instrumentedContent);
        return instrumentedContent;
    }
}
exports.ContentInstrumenter = ContentInstrumenter;
