// slnodejs/tsOutputs/build-scanner/diagnostic/contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectRootSource = exports.SourceMapType = exports.DiagnosticEventType = void 0;
var DiagnosticEventType;
(function (DiagnosticEventType) {
    DiagnosticEventType[DiagnosticEventType["FILE_SCANNED"] = 0] = "FILE_SCANNED";
    DiagnosticEventType[DiagnosticEventType["FILE_IGNORED"] = 1] = "FILE_IGNORED";
    DiagnosticEventType[DiagnosticEventType["FILE_SIGNATURE_CREATED"] = 2] = "FILE_SIGNATURE_CREATED";
    DiagnosticEventType[DiagnosticEventType["METHOD_ADDED_TO_SIGNATURE"] = 3] = "METHOD_ADDED_TO_SIGNATURE";
    DiagnosticEventType[DiagnosticEventType["FILE_SIGNATURE_CREATED_WITH_ERROR"] = 4] = "FILE_SIGNATURE_CREATED_WITH_ERROR";
    DiagnosticEventType[DiagnosticEventType["FILE_SOURCEMAP_IS_FOUND"] = 5] = "FILE_SOURCEMAP_IS_FOUND";
    DiagnosticEventType[DiagnosticEventType["FILE_SOURCEMAP_NOT_FOUND"] = 6] = "FILE_SOURCEMAP_NOT_FOUND";
    DiagnosticEventType[DiagnosticEventType["METHOD_WAS_MAPPED"] = 7] = "METHOD_WAS_MAPPED";
    DiagnosticEventType[DiagnosticEventType["ADD_MAPPED_METHOD"] = 8] = "ADD_MAPPED_METHOD";
    DiagnosticEventType[DiagnosticEventType["FILE_PATH_FIXED"] = 9] = "FILE_PATH_FIXED";
    DiagnosticEventType[DiagnosticEventType["METHOD_UNIQUE_ID_FIXED"] = 10] = "METHOD_UNIQUE_ID_FIXED";
    DiagnosticEventType[DiagnosticEventType["SOURCEMAP_CONSUMER"] = 11] = "SOURCEMAP_CONSUMER";
    DiagnosticEventType[DiagnosticEventType["PROJECT_ROOT_RESOLVED"] = 12] = "PROJECT_ROOT_RESOLVED";
    DiagnosticEventType[DiagnosticEventType["USED_PLUGINS"] = 13] = "USED_PLUGINS";
    DiagnosticEventType[DiagnosticEventType["MISSED_PLUGIN"] = 14] = "MISSED_PLUGIN";
    DiagnosticEventType[DiagnosticEventType["ADDED_PLUGIN"] = 15] = "ADDED_PLUGIN";
    DiagnosticEventType[DiagnosticEventType["ENRICHMENT_SKIPPED"] = 16] = "ENRICHMENT_SKIPPED";
    DiagnosticEventType[DiagnosticEventType["GENERETED_FILE_IGNORED"] = 17] = "GENERETED_FILE_IGNORED";
    DiagnosticEventType[DiagnosticEventType["AUTOGENERATED_METHOD"] = 18] = "AUTOGENERATED_METHOD";
    DiagnosticEventType[DiagnosticEventType["METHOD_ENRICHED"] = 19] = "METHOD_ENRICHED";
})(DiagnosticEventType = exports.DiagnosticEventType || (exports.DiagnosticEventType = {}));
var SourceMapType;
(function (SourceMapType) {
    SourceMapType[SourceMapType["NONE"] = 0] = "NONE";
    SourceMapType[SourceMapType["EMBEDDED"] = 1] = "EMBEDDED";
    SourceMapType[SourceMapType["LOCAL"] = 2] = "LOCAL";
    SourceMapType[SourceMapType["URL"] = 3] = "URL";
})(SourceMapType = exports.SourceMapType || (exports.SourceMapType = {}));
var ProjectRootSource;
(function (ProjectRootSource) {
    ProjectRootSource["COMMAND_ARG"] = "command-argument";
    ProjectRootSource["SCM"] = "scm";
    ProjectRootSource["CWD"] = "cwd";
})(ProjectRootSource = exports.ProjectRootSource || (exports.ProjectRootSource = {}));
