// slnodejs/tsOutputs/common/scm/none.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.None = void 0;
class None {
    constructor() {
    }
    detectCommitVersion(handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return 'N/A';
        });
    }
    getCommitOnlyHistory(historyLength, handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return [];
        });
    }
    getGitDiffs(refCommit, handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return [];
        });
    }
    getBranchHistory(handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                commitHistory: [],
                commitLog: [],
                commitToIndexMap: {}
            };
        });
    }
    getCommitsPerFile(handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return {};
        });
    }
    getRepositoryUrl(handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return '';
        });
    }
    getRootDirectory(handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return '';
        });
    }
    getFiles(handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return null;
        });
    }
    getOpenGitBranches(handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return [];
        });
    }
    getGitSubmodules(handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return [];
        });
    }
    getCurrentGitBranch(handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return '';
        });
    }
    getContributors(handlerParams, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                contributors: [],
                emailToIndexMap: {}
            };
        });
    }
}
exports.None = None;
