// slnodejs/tsOutputs/test-listener/entities/BasicConfigurationData.js
"use strict";
/**
 * Created by nadav on 12/03/2017.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicConfigurationData = void 0;
class BasicConfigurationData {
    get proxy() {
        return this._proxy;
    }
    set proxy(proxy) {
        this._proxy = proxy;
    }
    get appName() {
        return this._appName;
    }
    set appName(appName) {
        this._appName = appName;
    }
    get buildSessionId() {
        return this._buildSessionId;
    }
    set buildSessionId(buildSessionId) {
        this._buildSessionId = buildSessionId;
    }
    get buildSessionIdFile() {
        return this._buildSessionIdFile;
    }
    set buildSessionIdFile(buildSessionIdFile) {
        this._buildSessionIdFile = buildSessionIdFile;
    }
    get tokenFile() {
        return this._tokenFile;
    }
    set tokenFile(tokenFile) {
        this._tokenFile = tokenFile;
    }
    get token() {
        return this._token;
    }
    set token(token) {
        this._token = token;
    }
}
exports.BasicConfigurationData = BasicConfigurationData;
