// slnodejs/tsOutputs/common/footprints-process-v6/collector-footprints-buffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CollectorFootprintsBuffer = void 0;
const agent_events_contracts_1 = require("../agent-events/agent-events-contracts");
const contracts_1 = require("../contracts");
const footprints_buffer_1 = require("./footprints-buffer");
class CollectorFootprintsBuffer extends footprints_buffer_1.FootprintsBuffer {
    constructor(agentInstanceData, agentConfig, footprintsMapping) {
        super(agentInstanceData, agentConfig);
        this.meta.footprintsMapping =
            footprintsMapping === contracts_1.FootprintsMapping.SERVER
                ? footprintsMapping
                : contracts_1.FootprintsMapping.AGENT;
        this.meta.agentConfig = {
            agentId: this.agentInstanceData.agentId,
            appName: this.agentConfig.appName.value,
            build: this.agentConfig.build.value,
            branch: this.agentConfig.branch.value,
            buildSessionId: this.agentConfig.buildSessionId.value,
            labId: this.agentConfig.labId.value,
            agentVersion: this.agentInstanceData.agentVersion,
            agentType: agent_events_contracts_1.AgentTypes.BROWSER_AGENT
        };
    }
}
exports.CollectorFootprintsBuffer = CollectorFootprintsBuffer;
