// slnodejs/tsOutputs/common/footprints-process-v6/relative-path-resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelativePathResolver = void 0;
class RelativePathResolver {
    constructor(projectRoot) {
        this.absoluteToRelativePath = new Map();
        this.projectRoot = projectRoot;
    }
    getRelativePath(absolutePath) {
        if (!this.absoluteToRelativePath.get(absolutePath)) {
            absolutePath = this.adjustPathSlashes(absolutePath);
            let relativePath = absolutePath.replace(this.resolveProjectRoot(), '');
            if (relativePath.indexOf('/') === 0) {
                relativePath = relativePath.substring(1);
            }
            this.absoluteToRelativePath.set(absolutePath, relativePath);
        }
        return this.absoluteToRelativePath.get(absolutePath);
    }
    adjustPathSlashes(filePath) {
        return (filePath || '').replace(/\\/g, '/');
    }
    resolveProjectRoot() {
        return (this.projectRoot || process.cwd()).replace(/\\/g, '/');
    }
}
exports.RelativePathResolver = RelativePathResolver;
