// slnodejs/tsOutputs/common/cli-progress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliProgress = void 0;
const cliProgress = require("cli-progress");
class CliProgress {
    constructor(log) {
        this.log = log;
    }
    start({ total, startValue }) {
        var _a, _b;
        this.progress = new cliProgress.SingleBar({}, cliProgress.Presets.shades_classic);
        (_b = (_a = this.progress) === null || _a === void 0 ? void 0 : _a.start) === null || _b === void 0 ? void 0 : _b.call(_a, total, startValue);
    }
    update({ currentValue }) {
        var _a, _b;
        if (!this.progress)
            return this.log.warn(`Cli progress is triggered but wasn't started`);
        (_b = (_a = this.progress) === null || _a === void 0 ? void 0 : _a.update) === null || _b === void 0 ? void 0 : _b.call(_a, currentValue);
    }
    stop() {
        var _a, _b;
        if (!this.progress)
            return this.log.warn(`Cli progress is triggered but wasn't started`);
        (_b = (_a = this.progress) === null || _a === void 0 ? void 0 : _a.stop) === null || _b === void 0 ? void 0 : _b.call(_a);
    }
}
exports.CliProgress = CliProgress;
