// slnodejs/tsOutputs/common/source-map-reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceMapReader = void 0;
const ConvertSourceMap = require("convert-source-map");
const path = require("path");
const fs = require("fs");
const validation_utils_1 = require("./utils/validation-utils");
/**
 * Reads source map data from generated file, uses 'convert-source-map' packge to resolve.
 * If not found trying to search relevat map file under same directory as gem=nerated.
 */
class SourceMapReader {
    constructor(filename, logger) {
        validation_utils_1.ValidationUtils.verifyNotNullOrEmpty(filename, 'filename');
        validation_utils_1.ValidationUtils.verifyNotNullOrEmpty(logger, 'logger');
        this.filename = filename;
        this.logger = logger;
    }
    readSourceMapData(generatedContent) {
        if (!generatedContent) {
            this.logger.debug('\'generatedContent\' cannot be null or empty');
            return null;
        }
        this.logger.debug(`About to resolve source map data for '${this.filename}'`);
        const sourceMapData = ConvertSourceMap.fromSource(generatedContent) ||
            ConvertSourceMap.fromMapFileSource(generatedContent, path.dirname(this.filename));
        if (sourceMapData && sourceMapData.sourcemap) {
            this.logger.info(`Found source map data for '${this.filename}'`);
            return sourceMapData.sourcemap;
        }
        else {
            const mapFromFile = this.readFromFile();
            if (mapFromFile) {
                return mapFromFile;
            }
            else {
                if (this.isUsingSourceURL(generatedContent)) {
                    //TODO: Fix message
                    this.logger.info('Seems source map comment is in unsupported format...');
                }
                this.logger.debug(`Couldn't find source map data for '${this.filename}'`);
                return null;
            }
        }
    }
    readFromFile() {
        let sourceMap = null;
        const mapFileName = this.filename + SourceMapReader.MAP_FILE_EXTENSION;
        const isMapFileExists = fs.existsSync(mapFileName);
        if (!isMapFileExists) {
            this.logger.info(`File '${mapFileName}' not exists.`);
            return null;
        }
        this.logger.info(`Found map file '${mapFileName}'`);
        try {
            const rawMapData = fs.readFileSync(mapFileName).toString();
            sourceMap = JSON.parse(rawMapData);
            return sourceMap;
        }
        catch (e) {
            this.logger.error(`Could not parse source map data from '${path}`);
            this.logger.error(`Error: '${e}`);
            return null;
        }
    }
    /**
     *  The 'convert-source-map' detects only comments using 'sourceMappingURL' which is the new standard,
     some tools generate with comment 'sourceURL' which not supported
     */
    isUsingSourceURL(content) {
        const fileMatch = content.match(SourceMapReader.MAP_FILE_COMMENT_REGEX);
        const commentMatch = content.match(SourceMapReader.COMMENT_REGEX);
        return commentMatch !== null || fileMatch !== null;
    }
}
exports.SourceMapReader = SourceMapReader;
/*
    The following regex's based on regex used in 'convert-source-map' to detect source map.
    See https://github.com/thlorenz/convert-source-map/blob/master/index.js
 */
SourceMapReader.COMMENT_REGEX = /^\s*\/(?:\/|\*)[@#]\s+sourceURL=data:(?:application|text)\/json;(?:charset[:=]\S+?;)?base64,(?:.*)$/mg;
SourceMapReader.MAP_FILE_COMMENT_REGEX = /(?:\/\/[@#][ \t]+sourceURL=([^\s'"`]+?)[ \t]*$)|(?:\/\*[@#][ \t]+sourceURL=([^\*]+?)[ \t]*(?:\*\/){1}[ \t]*$)/mg;
SourceMapReader.MAP_FILE_EXTENSION = '.map';
