// slnodejs/tsOutputs/test-listener/coloring/listening-port.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListeningPort = void 0;
const http = require("http");
/**
 * Created by nadav on 19/03/2017.
 */
/**
 * This class creates an HTTP server that can be used in order to color the agent.
 * We'll probably need that when the client's uses something that is unsupported (ie, not HTTP).
 * The actual color will be set by the HttpColoring class.
 * */
class ListeningPort {
    setup(config, stateTracker, log) {
        this.setupListeningPort(config, log);
    }
    teardown() {
        this.stopListening();
    }
    setupListeningPort(config, log) {
        const httpListeningPort = config.httpListeningPort.value;
        if (httpListeningPort) {
            //Create a server
            this.server = http.createServer(this.handleRequest);
            //Lets start our server
            this.server.listen(httpListeningPort, function () {
                log.info('[HTTP LISTENER] Agent is listening on port ' + httpListeningPort + ']');
                try {
                    this.server.unref(); //Don't hold the process active just because this server is up.
                }
                catch (e) {
                }
            });
        }
    }
    handleRequest(request, response) {
        response.end();
    }
    stopListening() {
        if (this.server) {
            this.server.close();
        }
    }
}
exports.ListeningPort = ListeningPort;
