// slnodejs/tsOutputs/test-listener/queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Queue = void 0;
const events = require("events");
class Queue extends events.EventEmitter {
    constructor() {
        super(...arguments);
        this.queue = [];
        this.maxItemsInQueue = 1000;
    }
    requeue(items) {
        if (!items || !items.length)
            return;
        this.queue = items.concat(this.queue);
    }
    enqueue(item) {
        this.queue.push(item);
        if (this.maxItemsInQueue == this.queue.length) {
            this.emit('full');
        }
    }
    getQueueSize() {
        return this.queue.length;
    }
    dequeue(size) {
        if (size == 0)
            size = this.queue.length; //dequeue all
        if (this.queue.length == 0)
            return [];
        if (size > this.queue.length)
            size = this.queue.length;
        const dequeuedItems = this.queue.splice(0, size);
        return dequeuedItems;
    }
    clear() {
        this.queue = [];
    }
}
exports.Queue = Queue;
