// slnodejs/tsOutputs/test-listener/UserMessageHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserMessageHelper = void 0;
/**
 * Created by nadav on 12/03/2017.
 */
class UserMessageHelper {
    constructor(perfix, verbose) {
        this.perfix = perfix;
        this.verbose = verbose;
    }
    println(text, ...args) {
        let message = this.perfix + ' - ' + text;
        if (args && args.length > 0) {
            for (let i = 0; i < args.length; i++) {
                const arg = args[i];
                if (arg instanceof Error) {
                    message += arg.message + '\n' + arg.stack;
                }
                else if (typeof arg === 'object') {
                    message += JSON.stringify(arg);
                }
                else {
                    message = message + arg.toString();
                }
            }
        }
        console.log(message);
    }
    printWarn(text) {
        this.println('WARNING - ' + text);
    }
    printError(text, error) {
        this.println('ERROR - ' + text, error);
    }
    printDebug(text) {
        if (this.verbose) {
            this.println('DEBUG - ' + text);
        }
    }
}
exports.UserMessageHelper = UserMessageHelper;
